R/MSnExperiment-class.R

```diff
@@ -1,6 +1,26 @@
 #' @include hidden_aliases.R
 NULL
 
+#' validation function for MSnExperiment
+#'
+#' @author Johannes Rainer
+#'
+#' @noRd
+.validMSnExperiment <- function(object) {
+    msg <- NULL
+    if (length(object@files)) {
+        if (length(object@files) != nrow(object@sampleData))
+            msg <- c(msg, paste0("Number of files does not match the number",
+                                 " of rows of 'sampleNames'"))
+        if (is.null(object@backend))
+            msg <- c(msg, paste0("If files are present the backend can not be",
+                                 " NULL"))
```

Comment:
I like the way validity checking is done in S4Vectors. TLDR: a function for each slot/aspect that is checked.
https://github.com/Bioconductor/S4Vectors/blob/39bd2363b55a0f18ea65ef9011a823151d708e76/R/DataFrame-class.R#L88-L122

This may makes writing unit tests easier. What do you think?

Reply:
Looks nice - but isn't it too much of an overhead implementing one function for each possible failure? I think it makes mostly sense if one re-uses checks. Let's see how you implement your checks and then I'll eventually adapt 😉